core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -877,6 +985,26 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
     return builder.build();
   }
 
+  private static RowKey freshRowKey(int keyId, Schema schema, RowKey rowKey) {
+    RowKey.Builder builder = RowKey.builderFor(schema).withKeyId(keyId);
+
+    for (RowKeyField field : rowKey.fields()) {
+      // look up the name of the source field in the old schema to get the new schema's id
+      String columnName = rowKey.schema().findColumnName(field.sourceId());
+      Preconditions.checkNotNull(columnName,
+          "Cannot find column in the row key's schema. id: %s, schema: %s",
+          field.sourceId(), rowKey.schema());
+
+      // reassign all row keys with fresh column IDs.
+      Types.NestedField column = schema.findField(columnName);
+      Preconditions.checkNotNull(column,
+          "Cannot find column in the fresh schema. name: %s, schema: %s", columnName, schema);
```

Comment:
We don't have to do the nullable check here, because the `builder.addField(columnName)` will do the check inside its implementation.  (I'm sorry I did not describe this clearly in the last comment).